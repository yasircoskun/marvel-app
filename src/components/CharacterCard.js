import React from 'react';
import TabLoader from './TabLoader';

class CharacterCard extends React.Component {
  async componentDidMount() {
    document.querySelectorAll('#CharacterCard_' + this.props.characterID + ' .TabHeader li').forEach(elem => {
      elem.onclick = (e) => {
        e.target.className = 'active';
        console.log(this.props.characterID);
        document.querySelector('#CharacterCard_' + this.props.characterID + ' #' + e.target.innerText).style.display = 'flex';
        document.querySelectorAll('#CharacterCard_' + this.props.characterID + ' .TabHeader li').forEach(sibling => {
          if (sibling !== e.target) {
            sibling.className = '';
            document.querySelector('#CharacterCard_' + this.props.characterID + ' #' + sibling.innerText).style.display = 'none';
          }
        })
      }
    })
  }

  ucFirst(str) {
    return str.charAt(0).toUpperCase() + str.slice(1)
  }

  render() {
    return (
      <div id={'CharacterCard_' + this.props.characterID} className="CharacterCard">
        <div className='TabArea' key={'TabArea_' + this.props.characterID}>
          <ul className='TabHeader'>
            <li className='active'>General</li>
            {Object.keys(this.props.data).map((key) => {
              if (typeof this.props.data[key]['collectionURI'] !== "undefined") {
                return (<li>{this.ucFirst(key)}</li>);
              }
              return false;
            })}
          </ul>
          <div className='TabBody'>
            <TabLoader
              characterID={this.props.characterID}
              data={this.props.data}
            ></TabLoader>
          </div>
        </div>
      </div>
    );
  }
}

export default CharacterCard;
