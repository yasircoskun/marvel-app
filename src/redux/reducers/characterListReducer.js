import { createSlice } from '@reduxjs/toolkit'

const storageName = "characterList"

export const characterListSlice = createSlice({
  name: storageName,
  initialState: {
    value: getStateFromLocalStorage(storageName),
  },
  reducers: {
    concat: (state, action) => {
      state.value = [...state.value, ...action.payload]
      localStorage.setItem(storageName, JSON.stringify(state.value))
    },
  },
})

export const { concat } = characterListSlice.actions


/**
 * The selectFavs function is exported for use in components 
 * @param {any} state
 * @returns {any}
 */
export const selectCharacterList = () => getStateFromLocalStorage(storageName)

/**
 * Retrieves named state data from local storage.
 * @param {any} name
 * @returns {any}
 */
function getStateFromLocalStorage(name){
  let stateFromLocalStorage = localStorage.getItem(name);
  if(stateFromLocalStorage !== null){
    return JSON.parse(stateFromLocalStorage);
  }
  return []
}

export default characterListSlice.reducer
