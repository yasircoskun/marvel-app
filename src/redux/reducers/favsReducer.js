import { createSlice } from '@reduxjs/toolkit'

const storageName = "favs"

export const favsSlice = createSlice({
  name: storageName,
  initialState: {
    value: getStateFromLocalStorage(storageName),
  },
  reducers: {
    add: (state, action) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      // state.value += 1
      state.value[action.payload.name] = action.payload
      localStorage.setItem(storageName, JSON.stringify(state.value))
    },
    del: (state, action) => {
      delete state.value[action.payload.name]
      localStorage.setItem(storageName, JSON.stringify(state.value))
    },
  },
})

export const { add, del } = favsSlice.actions


/**
 * The selectFavs function is exported for use in components 
 * @param {any} state
 * @returns {any}
 */
export const selectFavs = () => getStateFromLocalStorage(storageName)

/**
 * Retrieves named state data from local storage.
 * @param {any} name
 * @returns {any}
 */
function getStateFromLocalStorage(name){
  let stateFromLocalStorage = localStorage.getItem(name);
  if(stateFromLocalStorage !== null){
    return JSON.parse(stateFromLocalStorage);
  }
  return {}
}

export default favsSlice.reducer
