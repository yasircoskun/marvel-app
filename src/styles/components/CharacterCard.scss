.CharacterCard {
  width: 60vw;
  background-color: #ccc;
  margin: 0.4rem 20vw;
  padding: 2rem 1.25rem;
  border-radius: 1.5rem;
  display: flex;
  text-align: left;
  .Image,
  .TabArea {
    display: inline-block;
  }
  .TabArea {
    width: 100%;
    li {
      display: inline-block;
      position: static;
      padding: 0.8rem 0.5rem 0.3rem 0.5rem;
      border: 2px solid $marvel_red;
      min-width: 4rem;
      cursor: pointer;
      border-radius: 1.5rem 1.5rem 0 0;
      text-align: center;
      background-color: $marvel_red_2;
      color: white;
    }
    $number: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10;
    @each $num in $number {
      li:nth-child(#{$num}) {
        transform: translate($num * -0.6rem, 0px);
      }
    }
    li.active {
      z-index: 999;
      position: relative;
    }
    .TabHeader {
      margin: 0;
      padding: 0 1.4rem;
    }
    .TabBody {
      border: 2px solid $marvel_red;
      margin: 0;
      border-radius: 1rem;
      .Tab {
        display: none;
        height: 324px;
        overflow: hidden;
        .Info .label {
          font-weight: bold;
        }
        .Info {
          width: 100%;
          position: relative;
          overflow: auto;
          height: 324px;
          .label,
          .data {
            display: inline-block;
            margin: 0.2em 1em;
          }
        }
        .Image img {
          border-radius: 1rem 0 0 1rem;
          height: 324px;
        }
        .button {
          display: inline-block;
          padding: 0.7rem 0.5rem 0.3rem 0.5rem;
          border: 2px solid $marvel_red_2;
          min-width: 4rem;
          cursor: pointer;
          text-align: center;
          background-color: $marvel_red_2;
          color: white;
          margin: 0.5rem;
          a {
            color: inherit;
          }
        }
      }
      .Tab:first-child {
        display: flex;
        height: 324px;
      }
    }
  }
}

.Object {
  padding: 0 0.5rem;
}

.ArrayContent,
.ObjectContent {
  padding: 0.5rem;
  border: 1px solid $marvel_red_2;
}

.Thumbnail {
  text-align: center;
  height: 324px;
  img {
    height: 324px;
  }
}

.ScrollList {
  position: relative;
  width: 100%;
  .ScrollListElement {
    display: flex;
  }
  .ScrollBtn {
    position: absolute;
    right: 0.5rem;
    bottom: 0.5rem;
  }
}

@media (max-width: 768px) {
  .CharacterCard {
    width: 88vw;
    background-color: #ccc;
    margin: 0.4rem 1vw;
    padding: 2rem 1.25rem;
    border-radius: 1.5rem;
    display: flex;
    text-align: left;
  }
  .CharacterCard .TabArea li {
    padding: 0.2rem 0.8rem 0rem 0.8rem;
    min-width: 2rem;
    border-radius: 4rem 4rem 0 0;
    font-size: x-small;
  }
  .CharacterCard .TabArea .TabBody .Tab:first-child {
    display: flex;
    height: 90vh;
  }
  .CharacterCard .TabArea .TabBody .Tab {
    display: none;
    height: 90vh;
  }
  .ScrollListElement {
    display: block;
    height: 90vh;
  }
  .Thumbnail {
    text-align: center;
    height: auto;
  }
  .CharacterCard .Image {
    display: block;
  }
  .Image img {
    border-radius: 1rem 1rem 0 0;
    height: auto;
    width: 100%;
  }
  .Info {
    width: 100%;
    position: relative;
    overflow: auto;
    //min-height: calc(~"90vh - 500px")
  }
}
