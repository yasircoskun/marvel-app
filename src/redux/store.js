import { configureStore } from '@reduxjs/toolkit';
import favsReducer from './reducers/favsReducer';
import characterListReducer from './reducers/characterListReducer';

export default configureStore({
  reducer: {
    favs: favsReducer,
    characterList: characterListReducer,
  },
});
