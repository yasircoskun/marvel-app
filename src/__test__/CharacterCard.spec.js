/* Dependencies  */
import React from 'react'
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer'
import { BrowserRouter as Router } from "react-router-dom";
import { add, del } from '../redux/reducers/favsReducer';
import store from '../redux/store';

/* Components */
import CharacterCard from '../components/CharacterCard'

const CharacterData = {
  "name": "Apocalypse",
  "characterID": 1009156,
  "data": {
    "id": 1009156,
    "name": "Apocalypse",
    "description": "",
    "modified": "2014-05-28T12:41:41-0400",
    "thumbnail": {
      "path": "http://i.annihil.us/u/prod/marvel/i/mg/f/e0/526166076a1d0",
      "extension": "jpg"
    },
    "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009156",
    "comics": {
      "available": 123,
      "collectionURI": "http://gateway.marvel.com/v1/public/characters/1009156/comics",
      "items": ["..."],
      "returned": 20
    },
    "series": {
      "available": 56,
      "collectionURI": "http://gateway.marvel.com/v1/public/characters/1009156/series",
      "items": ["..."],
      "returned": 20
    },
    "stories": {
      "available": 124,
      "collectionURI": "http://gateway.marvel.com/v1/public/characters/1009156/stories",
      "items": ["..."],
      "returned": 20
    },
    "events": {
      "available": 6,
      "collectionURI": "http://gateway.marvel.com/v1/public/characters/1009156/events",
      "items": ["..."],
      "returned": 6
    },
    "urls": ["..."]
  }
}

describe('CharacterCard', () => {
  it('Case 1: if props is empty', () => {
    const renderedValue =  renderer.create(<Provider store={store}><CharacterCard /></Provider>).toJSON()
    expect(renderedValue).toMatchSnapshot();
  })

  it('Case 2: if props contain only valid name', () => {
    const renderedValue =  renderer.create(
      <Provider store={store}>
        <CharacterCard name="Michael Carleone"/>
      </Provider>
    ).toJSON()
    expect(renderedValue).toMatchSnapshot();
  })

  it('Case 3: if props contain valid name and characterID', () => {
    const renderedValue =  renderer.create(
      <Provider store={store}>
        <CharacterCard name="Michael Carleone" characterID={1}/>
      </Provider>
    ).toJSON()
    expect(renderedValue).toMatchSnapshot();
  })

  it('Case 4: if props contain valid name, characterID and data', () => {
    const renderedValue =  renderer.create(
      <Provider store={store}>
        <Router>
          <CharacterCard name="Michael Carleone" characterID={1} data={{thumbnail: {extension: "", path: "1"}}}/>
        </Router>
      </Provider>
    ).toJSON()
    expect(renderedValue).toMatchSnapshot();
  })

  it('Case 5: if props contain valid props and in the favorites list (store)', () => {
    store.dispatch(add(CharacterData))
    const renderedValue = renderer.create(
      <Provider store={store}>
        <Router>
          <CharacterCard {...CharacterData}/>
        </Router>
      </Provider>
    ).toJSON()
    expect(renderedValue).toMatchSnapshot();
  })

  it('Case 6: if props contain valid props and not in the favorites list (store)', () => {
    store.dispatch(add(CharacterData))
    store.dispatch(del(CharacterData))
    const renderedValue = renderer.create(
      <Provider store={store}>
        <Router>
          <CharacterCard {...CharacterData}/>
        </Router>
      </Provider>
    ).toJSON()
    expect(renderedValue).toMatchSnapshot();
  })
})