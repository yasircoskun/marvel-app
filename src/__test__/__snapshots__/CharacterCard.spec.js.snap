// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard Case 1: if props is empty 1`] = `"Check Developer Tools Console in your browser"`;

exports[`CharacterCard Case 2: if props contain only valid name 1`] = `"Check Developer Tools Console in your browser"`;

exports[`CharacterCard Case 3: if props contain valid name and characterID 1`] = `"Check Developer Tools Console in your browser"`;

exports[`CharacterCard Case 4: if props contain valid name, characterID and data 1`] = `
<div
  className="CharacterCard"
  id={1}
>
  <a
    href="/comics/1"
    onClick={[Function]}
  >
    <div
      className="Thumbnail"
    >
      <img
        alt="..."
        src="1/portrait_incredible."
      />
    </div>
    <div
      className="CardTitle"
    >
      <h4>
        Michael Carleone
      </h4>
    </div>
  </a>
  <div
    className="HeartIcon"
    onClick={[Function]}
  >
    <span
      className=""
    >
      ü§ç
    </span>
  </div>
</div>
`;

exports[`CharacterCard Case 5: if props contain valid props and in the favorites list (store) 1`] = `
<div
  className="CharacterCard"
  id={1009156}
>
  <a
    href="/comics/1009156"
    onClick={[Function]}
  >
    <div
      className="Thumbnail"
    >
      <img
        alt="..."
        src="http://i.annihil.us/u/prod/marvel/i/mg/f/e0/526166076a1d0/portrait_incredible.jpg"
      />
    </div>
    <div
      className="CardTitle"
    >
      <h4>
        Apocalypse
      </h4>
    </div>
  </a>
  <div
    className="HeartIcon"
    onClick={[Function]}
  >
    <span
      className="faved"
    >
      ‚ù§Ô∏è
    </span>
  </div>
</div>
`;

exports[`CharacterCard Case 6: if props contain valid props and not in the favorites list (store) 1`] = `
<div
  className="CharacterCard"
  id={1009156}
>
  <a
    href="/comics/1009156"
    onClick={[Function]}
  >
    <div
      className="Thumbnail"
    >
      <img
        alt="..."
        src="http://i.annihil.us/u/prod/marvel/i/mg/f/e0/526166076a1d0/portrait_incredible.jpg"
      />
    </div>
    <div
      className="CardTitle"
    >
      <h4>
        Apocalypse
      </h4>
    </div>
  </a>
  <div
    className="HeartIcon"
    onClick={[Function]}
  >
    <span
      className=""
    >
      ü§ç
    </span>
  </div>
</div>
`;
